flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointMetadataOutputStream.java

```diff
@@ -62,7 +64,15 @@ public FsCheckpointMetadataOutputStream(
         this.metadataFilePath = checkNotNull(metadataFilePath);
         this.exclusiveCheckpointDir = checkNotNull(exclusiveCheckpointDir);
 
-        this.out = fileSystem.create(metadataFilePath, WriteMode.NO_OVERWRITE);
+        try {
+            RecoverableWriter recoverableWriter = fileSystem.createRecoverableWriter();
```

Comment:
I have a bit concern on the compatibility when the `RecoverableWriter` could not be created: currently it only falls back to the normal output stream when the FileSystem throws UnsupportedOperationException. However, since users may have customized FileSystem implementation that changes this behavior for unsupported cases. 

Might we fallback to the normal output stream with warnings whenever the `RecoverableWriter` failed to create? like

```
RecoverableWriter recoverableWriter = null;
try {
    recoverableWriter = ...
} catch (Throwable e) {
   Log.Warn(...)
}

if (recoverableWriter != null) {
    // use recoverable writer
} else {
   // use normal output stream
}

```

Reply:
Make sense.